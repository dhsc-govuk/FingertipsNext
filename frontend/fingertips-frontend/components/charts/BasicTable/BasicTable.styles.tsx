import styled from 'styled-components';
import { Table } from 'govuk-react';

export const StyledTable = styled(Table)({
  marginBottom: '1rem',
});
