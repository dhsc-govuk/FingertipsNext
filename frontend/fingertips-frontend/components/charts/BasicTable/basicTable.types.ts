import type { HealthDataPointTrendEnum } from '@/generated-sources/ft-api-client';

export enum BasicTableEnum {
  Indicator = 'Indicator',
  Period = 'Period',
  Count = 'Count',
  ValueUnit = 'Value unit',
  Value = 'Value',
  RecentTrend = 'Recent trend',
}

export interface BasicTableData {
  indicatorId?: number;
  indicatorName?: string;
  areaName?: string;
  areaCode?: string;
  period?: string;
  unitLabel?: string;
  count?: number;
  value?: number;
  trend?: HealthDataPointTrendEnum;
}
