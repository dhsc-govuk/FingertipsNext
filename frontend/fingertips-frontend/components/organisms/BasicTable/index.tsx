'use client';

import {
  type HealthDataPoint,
  HealthDataPointTrendEnum,
} from '@/generated-sources/ft-api-client';
import { Table } from 'govuk-react';
import React from 'react';
import {
  CheckValueInTableCell,
  FormatNumberInTableCell,
} from '@/components/molecules/CheckValueInTableCell';
import { TrendTag } from '@/components/molecules/TrendTag';
import { StyledCenterTrendColumn } from '@/lib/tableHelpers';

export enum BasicTableEnum {
  Indicator = 'Indicator',
  Period = 'Period',
  Count = 'Count',
  ValueUnit = 'Value unit',
  Value = 'Value',
  RecentTrend = 'Recent trend',
}

export interface BasicTableData {
  indicatorId?: number;
  indicatorName?: string;
  period?: string;
  latestEnglandHealthData?: HealthDataPoint;
  unitLabel?: string;
}

interface BasicTableProps {
  indicatorData: BasicTableData[];
  areaName: string;
}

export function BasicTable({
  indicatorData,
  areaName,
}: Readonly<BasicTableProps>) {
  return (
    <div data-testid={'basicTable-component'}>
      <Table
        head={
          <React.Fragment>
            <Table.Row>
              <Table.CellHeader colSpan={6} style={{paddingLeft: '10px'}}>{areaName}</Table.CellHeader>
            </Table.Row>

            <Table.Row>
              <Table.CellHeader style={{ verticalAlign: 'top', paddingLeft: '10px' }}>
                {BasicTableEnum.Indicator}
              </Table.CellHeader>
              <Table.CellHeader style={{ verticalAlign: 'top' }}>
                {BasicTableEnum.Period}
              </Table.CellHeader>
              <Table.CellHeader
                style={{ verticalAlign: 'top', textAlign: 'right' }}
              >
                {BasicTableEnum.Count}
              </Table.CellHeader>
              <Table.CellHeader
                style={{
                  verticalAlign: 'top',
                  textAlign: 'right',
                }}
              >
                {BasicTableEnum.ValueUnit}
              </Table.CellHeader>
              <Table.CellHeader
                style={{ verticalAlign: 'top', textAlign: 'right' }}
              >
                {BasicTableEnum.Value}
              </Table.CellHeader>
              <Table.CellHeader
                style={{ verticalAlign: 'top', textAlign: 'center' }}
              >
                {BasicTableEnum.RecentTrend}
              </Table.CellHeader>
            </Table.Row>
          </React.Fragment>
        }
      >
        {indicatorData.map((item) => (
          <Table.Row key={item.indicatorId}>
            <CheckValueInTableCell value={item?.indicatorName} style={{ paddingLeft: '10px' }} />
            <CheckValueInTableCell
              value={item?.period}
              style={{ textAlign: 'right' }}
            />
            <FormatNumberInTableCell
              value={item?.latestEnglandHealthData?.count}
              numberStyle={'whole'}
              style={{ textAlign: 'right' }}
            />
            <CheckValueInTableCell
              value={item?.unitLabel}
              style={{ textAlign: 'right' }}
            />
            <FormatNumberInTableCell
              value={item?.latestEnglandHealthData?.value}
              style={{ textAlign: 'right' }}
            />
            <StyledCenterTrendColumn>
              <TrendTag
                trendFromResponse={
                  item?.latestEnglandHealthData?.trend ??
                  HealthDataPointTrendEnum.CannotBeCalculated
                }
              />
            </StyledCenterTrendColumn>
          </Table.Row>
        ))}
      </Table>
    </div>
  );
}
