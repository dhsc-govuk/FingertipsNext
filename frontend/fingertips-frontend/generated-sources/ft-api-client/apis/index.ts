/* tslint:disable */
/* eslint-disable */
export * from './AreasApi';
export * from './BatchesApi';
export * from './IndicatorsApi';
export * from './SystemApi';
export * from './UserApi';
