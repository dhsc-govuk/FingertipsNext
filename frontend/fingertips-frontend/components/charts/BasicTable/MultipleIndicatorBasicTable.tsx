import { BasicTable } from '@/components/charts/BasicTable/BasicTable';
import { H3 } from 'govuk-react';
import { StyleChartWrapper } from '@/components/styles/viewPlotStyles/styleChartWrapper';
import { useMultipleIndicatorBasicTableData } from '@/components/charts/BasicTable/hooks/useMultipleIndicatorBasicTableData';

export function MultipleIndicatorBasicTable() {
  const tableData = useMultipleIndicatorBasicTableData();

  if (!tableData || tableData.length <= 1) return null;

  return (
    <StyleChartWrapper>
      <H3>Indicator segmentations overview</H3>
      <BasicTable tableData={tableData} />
    </StyleChartWrapper>
  );
}
