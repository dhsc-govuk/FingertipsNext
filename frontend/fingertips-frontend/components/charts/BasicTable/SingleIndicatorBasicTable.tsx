import { BasicTable } from '@/components/charts/BasicTable/BasicTable';
import { H3 } from 'govuk-react';
import { StyleChartWrapper } from '@/components/styles/viewPlotStyles/styleChartWrapper';
import { useSingleIndicatorBasicTableData } from '@/components/charts/BasicTable/hooks/useSingleIndicatorBasicTableData';

export function SingleIndicatorBasicTable() {
  const tableData = useSingleIndicatorBasicTableData();

  if (!tableData || tableData.length <= 1) return null;

  return (
    <StyleChartWrapper>
      <H3>Indicator segmentations overview</H3>
      <BasicTable
        tableData={tableData}
        id={'singleIndicatorBasicTable'}
        title={'Segmentation overview of selected indicator'}
      />
    </StyleChartWrapper>
  );
}
