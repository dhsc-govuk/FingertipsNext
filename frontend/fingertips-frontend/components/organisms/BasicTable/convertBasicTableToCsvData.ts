import { CsvData, CsvRow } from '@/lib/downloadHelpers/convertToCsv';
import { BasicTableData } from '.';
import { CsvHeader } from '@/components/molecules/Export/export.types';
import { HealthDataPointTrendEnum } from '@/generated-sources/ft-api-client';
import { areaCodeForEngland } from '@/lib/chartHelpers/constants';

export const convertBasicTableToCsvData = (
  indicatorData: BasicTableData[]
): CsvData => {
  const header: CsvRow = [
    CsvHeader.IndicatorId,
    CsvHeader.IndicatorName,
    CsvHeader.Period,
    CsvHeader.Area,
    CsvHeader.AreaCode,
    CsvHeader.RecentTrend,
    CsvHeader.Count,
    CsvHeader.ValueUnit,
    CsvHeader.Value,
  ];

  const csvData: CsvData = [header];

  indicatorData.forEach((indicator) => {
    const csvRow: CsvRow = [
      indicator.indicatorId,
      indicator.indicatorName,
      indicator.period,
      'England',
      areaCodeForEngland,
      indicator.latestEnglandHealthData?.trend ??
        HealthDataPointTrendEnum.CannotBeCalculated,
      indicator.latestEnglandHealthData?.count,
      indicator.unitLabel,
      indicator.latestEnglandHealthData?.value,
    ];

    csvData.push(csvRow);
  });

  return csvData;
};
