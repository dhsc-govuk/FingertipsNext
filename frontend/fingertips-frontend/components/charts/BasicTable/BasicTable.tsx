'use client';

import { HealthDataPointTrendEnum } from '@/generated-sources/ft-api-client';
import { Table } from 'govuk-react';
import React from 'react';
import {
  CheckValueInTableCell,
  FormatNumberInTableCell,
} from '@/components/molecules/CheckValueInTableCell';
import { TrendTag } from '@/components/molecules/TrendTag';
import { StyledCenterTableCell } from '@/lib/tableHelpers';
import { ExportOptionsButton } from '@/components/molecules/Export/ExportOptionsButton';
import { convertBasicTableToCsvData } from './helpers/convertBasicTableToCsvData';
import { ExportCopyright } from '@/components/molecules/Export/ExportCopyright';
import { ExportOnlyWrapper } from '@/components/molecules/Export/ExportOnlyWrapper';
import { ChartTitle } from '@/components/atoms/ChartTitle/ChartTitle';
import { ContainerWithOutline } from '@/components/atoms/ContainerWithOutline/ContainerWithOutline';
import {
  BasicTableData,
  BasicTableEnum,
} from '@/components/charts/BasicTable/basicTable.types';
import { StyledTable } from '@/components/charts/BasicTable/BasicTable.styles';

interface BasicTableProps {
  tableData: BasicTableData[];
  title?: string;
}

export function BasicTable({
  tableData,
  title = 'Overview of selected indicators',
}: Readonly<BasicTableProps>) {
  const areaName = tableData.at(0)?.areaName;
  const csvData = convertBasicTableToCsvData(tableData);
  const id = 'basic-table';
  return (
    <div data-testid={`${id}-component`}>
      <ContainerWithOutline>
        <div id={id}>
          <ChartTitle>{title}</ChartTitle>
          <StyledTable
            head={
              <React.Fragment>
                <Table.Row>
                  <Table.CellHeader colSpan={6} style={{ paddingLeft: '10px' }}>
                    {areaName}
                  </Table.CellHeader>
                </Table.Row>

                <Table.Row>
                  <Table.CellHeader
                    style={{ verticalAlign: 'top', paddingLeft: '10px' }}
                  >
                    {BasicTableEnum.Indicator}
                  </Table.CellHeader>
                  <Table.CellHeader style={{ verticalAlign: 'top' }}>
                    {BasicTableEnum.Period}
                  </Table.CellHeader>
                  <Table.CellHeader
                    style={{ verticalAlign: 'top', textAlign: 'right' }}
                  >
                    {BasicTableEnum.Count}
                  </Table.CellHeader>
                  <Table.CellHeader
                    style={{
                      verticalAlign: 'top',
                      textAlign: 'left',
                    }}
                  >
                    {BasicTableEnum.ValueUnit}
                  </Table.CellHeader>
                  <Table.CellHeader
                    style={{ verticalAlign: 'top', textAlign: 'right' }}
                  >
                    {BasicTableEnum.Value}
                  </Table.CellHeader>
                  <Table.CellHeader
                    style={{ verticalAlign: 'top', textAlign: 'center' }}
                  >
                    {BasicTableEnum.RecentTrend}
                  </Table.CellHeader>
                </Table.Row>
              </React.Fragment>
            }
          >
            {tableData.map((item) => (
              <Table.Row key={item.indicatorName}>
                <CheckValueInTableCell
                  value={item?.indicatorName}
                  style={{ paddingLeft: '10px' }}
                />
                <CheckValueInTableCell
                  value={item?.period}
                  style={{ textAlign: 'right' }}
                />
                <FormatNumberInTableCell
                  value={item?.count}
                  numberStyle={'whole'}
                  style={{ textAlign: 'right' }}
                />
                <CheckValueInTableCell
                  value={item?.unitLabel}
                  style={{ textAlign: 'left' }}
                />
                <FormatNumberInTableCell
                  value={item?.value}
                  style={{ textAlign: 'right' }}
                />
                <StyledCenterTableCell>
                  <TrendTag
                    trendFromResponse={
                      item?.trend ?? HealthDataPointTrendEnum.CannotBeCalculated
                    }
                  />
                </StyledCenterTableCell>
              </Table.Row>
            ))}
          </StyledTable>
          <ExportOnlyWrapper>
            <ExportCopyright />
          </ExportOnlyWrapper>
        </div>
        <ExportOptionsButton targetId={id} csvData={csvData} />
      </ContainerWithOutline>
    </div>
  );
}
