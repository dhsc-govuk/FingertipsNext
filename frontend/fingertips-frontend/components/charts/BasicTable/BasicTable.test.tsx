import { render, screen } from '@testing-library/react';
import { BasicTable } from '@/components/charts/BasicTable/BasicTable';
import { BasicTableData } from '@/components/charts/BasicTable/basicTable.types';
import { HealthDataPointTrendEnum } from '@/generated-sources/ft-api-client';

const mockTableData: BasicTableData[] = [
  {
    indicatorId: 1,
    indicatorName: ' ',
    period: '2008',
    count: 389,
    value: 278.29134,
    trend: HealthDataPointTrendEnum.NotYetCalculated,
    unitLabel: '',
  },
  {
    indicatorId: 2,
    indicatorName: 'no data indicator',
    period: '',
    unitLabel: '',
  },
];

describe('BasicTable', () => {
  beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date('2024-12-25T12:00:00Z'));
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  it('should include a title', () => {
    render(<BasicTable tableData={mockTableData} />);
    const title = screen.getByRole('heading', { level: 4 });
    expect(title).toHaveTextContent('Overview of selected indicators');
  });

  it('should match snapshot', () => {
    render(<BasicTable tableData={mockTableData} />);
    const table = screen.getByRole('table');
    expect(table).toMatchSnapshot();
  });

  it('should render BasicTable component', () => {
    render(<BasicTable tableData={mockTableData} />);
    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.getByTestId('basic-table-component')).toBeInTheDocument();
  });

  it('should always display indicator name when no health data for the indicator is present', () => {
    render(<BasicTable tableData={mockTableData} />);

    expect(screen.getAllByRole('row')[3]).toHaveTextContent(
      'no data indicator'
    );
    expect(screen.getAllByRole('cell')[3]).toHaveTextContent('X');
  });

  it('should display X in the table cell if there is no value', () => {
    render(<BasicTable tableData={mockTableData} />);

    const noValueCells = screen.getAllByText('X');
    expect(noValueCells).toHaveLength(5);
  });

  it('should display the correct aria label when then is no value', () => {
    render(<BasicTable tableData={mockTableData} />);

    const noValueCells = screen.getAllByLabelText('Not compared');
    expect(noValueCells).toHaveLength(5);
  });

  it('should render the correct trend label for the data in the table', () => {
    render(<BasicTable tableData={mockTableData} />);

    const trendTags = screen.getAllByTestId('trend-tag-component');
    expect(trendTags).toHaveLength(2);
    expect(trendTags[0].textContent).toEqual('No recent trend data available');
    expect(trendTags[1].textContent).toEqual('No recent trend data available');
  });
});
