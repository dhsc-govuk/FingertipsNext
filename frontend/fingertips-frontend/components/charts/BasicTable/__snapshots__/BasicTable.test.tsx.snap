// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BasicTable > should match snapshot 1`] = `
<table
  class="src__StyledTable-sc-16s660v-0 duOufa sc-dtBdUo ekpRUm"
>
  <thead
    class="TableHeader-sc-1n1nw4b-0"
  >
    <tr
      class="Row-sc-1y8eqfp-0"
    >
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        colspan="6"
        style="padding-left: 10px;"
      />
    </tr>
    <tr
      class="Row-sc-1y8eqfp-0"
    >
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        style="vertical-align: top; padding-left: 10px;"
      >
        Indicator
      </th>
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        style="vertical-align: top;"
      >
        Period
      </th>
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        style="vertical-align: top; text-align: right;"
      >
        Count
      </th>
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        style="vertical-align: top; text-align: left;"
      >
        Value unit
      </th>
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        style="vertical-align: top; text-align: right;"
      >
        Value
      </th>
      <th
        class="Cell-sc-msv0l6-0 iftjBc"
        style="vertical-align: top; text-align: center;"
      >
        Recent trend
      </th>
    </tr>
  </thead>
  <tbody
    class="TableBody-sc-1qqarm8-0"
  >
    <tr
      class="Row-sc-1y8eqfp-0"
    >
      <td
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="padding-left: 10px;"
      >
         
      </td>
      <td
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: right;"
      >
        2008
      </td>
      <td
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: right;"
      >
        389
      </td>
      <td
        aria-label="Not compared"
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: left;"
      >
        X
      </td>
      <td
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: right;"
      >
        278.3
      </td>
      <td
        class="Cell-sc-msv0l6-0 sc-cwHptR gaSlHc itmUVF"
      >
        <div
          data-testid="trendTag-container"
        >
          <strong
            class="src__Tag-sc-10wvgvx-0 sc-gEvEer bujMNx dZEcNR"
          >
            <div
              class="sc-aXZVg dOZydJ"
              data-testid="trend-tag-component"
            >
              <div>
                <div
                  class="src__StyledParagraph-sc-sgc9po-0 qWnWb sc-eqUAAy crFqhO"
                >
                  <p>
                    No recent trend data available
                  </p>
                </div>
              </div>
            </div>
          </strong>
        </div>
      </td>
    </tr>
    <tr
      class="Row-sc-1y8eqfp-0"
    >
      <td
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="padding-left: 10px;"
      >
        no data indicator
      </td>
      <td
        aria-label="Not compared"
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: right;"
      >
        X
      </td>
      <td
        aria-label="Not compared"
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: right;"
      >
        X
      </td>
      <td
        aria-label="Not compared"
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: left;"
      >
        X
      </td>
      <td
        aria-label="Not compared"
        class="Cell-sc-msv0l6-0 gaSlHc"
        style="text-align: right;"
      >
        X
      </td>
      <td
        class="Cell-sc-msv0l6-0 sc-cwHptR gaSlHc itmUVF"
      >
        <div
          data-testid="trendTag-container"
        >
          <strong
            class="src__Tag-sc-10wvgvx-0 sc-gEvEer bujMNx dZEcNR"
          >
            <div
              class="sc-aXZVg dOZydJ"
              data-testid="trend-tag-component"
            >
              <div>
                <div
                  class="src__StyledParagraph-sc-sgc9po-0 qWnWb sc-eqUAAy crFqhO"
                >
                  <p>
                    No recent trend data available
                  </p>
                </div>
              </div>
            </div>
          </strong>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
